overlord/patch/patch4.go

```diff
@@ -0,0 +1,171 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patch
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+func init() {
+	patches[4] = patch4
+}
+
+type patch4T struct{} // for namespacing of the helpers
+
+func (patch4T) snapSetupAndState(task *state.Task) (*snapstate.SnapSetup, *snapstate.SnapState, error) {
+	var snapst snapstate.SnapState
+
+	ss, err := snapstate.TaskSnapSetup(task)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot get snap setup from task %s (%s): %v", task.ID(), task.Kind(), err)
+	}
+
+	err = snapstate.Get(task.State(), ss.Name(), &snapst)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot get state for snap %q: %v", ss.Name(), err)
+	}
+
+	return ss, &snapst, err
+}
+
+func (patch4T) get(task *state.Task, key string, passThroughMissing bool, value interface{}) error {
+	err := task.Get(key, value)
+	if err == nil || (passThroughMissing && err == state.ErrNoState) {
+		return err
+	}
+	change := task.Change()
+
+	return fmt.Errorf("cannot get %q from task %s (%s) of change %s (%s): %v",
+		key, task.ID(), task.Kind(), change.ID(), change.Kind(), err)
+}
+
+func (p4 patch4T) addCleanup(task *state.Task) error {
+	// NOTE we could check for the status of the change itself, but
+	// copy-snap-data is the one creating the trash, so if it's run there's
+	// no sense in fiddling with the change.
+	if task.Status().Ready() {
+		return nil
+	}
+
+	ss, snapst, err := p4.snapSetupAndState(task)
+	if err != nil {
+		return err
+	}
+	if !snapst.HasCurrent() {
```

Comment:
would be best if didn't add new code using HasCurrent, at least for the snapstate.Get case it doesn't make sense anymore and I would like to remove it at some point